### PR TITLE
[core] GidIO change to avoid memory problems when having multiple instances (due to different template arguments)

### PR DESCRIPTION
**Pull request by maceligueta**
A base class has been created, which holds the static variable 'msLiveInstances'. This base class is not a template, so this variable is unique in memory.
Solves #4337 .

### CODE REVIEW

**Comment by philbucher** [comment]
does this solve the problems you have?

File: kratos/sources/gid_io.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/gid_io.h"
+
+namespace Kratos
+{
+    int GiDIOBase::msLiveInstances = 0;
```
Comment (philbucher) [inline kratos/sources/gid_io.cpp]: why does this need to be in the cpp?

Comment (maceligueta) [reply kratos/sources/gid_io.cpp]: Hi @philbucher, I did not manage to add it in the .h file. As it is not const, it was not compiling/linking.

Comment (maceligueta) [reply kratos/sources/gid_io.cpp]: After trying several combinations I googled the problem and found this: https://stackoverflow.com/questions/185844/how-to-initialize-private-static-members-in-c

**Comment by philbucher** [comment]
Looking at the Appveyor output you probably have to move the static var to protected

**Comment by philbucher** [comment]
> Looking at the Appveyor output you probably have to move the static var to protected

hm did not work :/

**Comment by maceligueta** [comment]
@philbucher This change actually solved my problem in linux. I just tried with the 'protected' version. The segmentation fault no longer appears at the end.

File: kratos/sources/gid_io.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/gid_io.h"
+
+namespace Kratos
+{
+    int GidIOBase::msLiveInstances = 0;
```
Comment (philbucher) [inline kratos/sources/gid_io.cpp:22]: how does this work if you have multiple instances or multiple classes deriving from the `GidIOBase`?

Did you try with the one where you have problems?

Comment (maceligueta) [reply kratos/sources/gid_io.cpp:22]: Hi @philbucher . The plan is that there's only one copy of `msLiveInstances` , for all instances of GidIO, even with different template arguments (different classes deriving from GidIOBase). The fact of `msLiveInstances` being `static` and belonging to GidIOBase makes it unique. 
The problem might come when GidIO is derived, if constructors of GidIODerived do not call the constructor of GidIO...
Other problems can appear it two instances modify 'msLiveInstances' simultaneously from different threads.

Comment (maceligueta) [reply kratos/sources/gid_io.cpp:22]: I tried the implementation last week (linux), and the problem had gone. I will run it now again.

Comment (maceligueta) [reply kratos/sources/gid_io.cpp:22]: It runs fine.

**Review by jginternational — APPROVED** [review]

**Comment by maceligueta** [comment]
This is working for @jginternational  in windows and for me in Linux. Apart from this, it is (unexpectedly) solving extra problems (#5868). So I am merging.

**Pull request merged**